pkg/rbd/controllerserver.go

```diff
@@ -278,6 +279,45 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
 	return &csi.DeleteVolumeResponse{}, nil
 }
 
+// ListVolumes returns a list of volumes stored in memory
+func (cs *ControllerServer) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (*csi.ListVolumesResponse, error) {
+
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_LIST_VOLUMES); err != nil {
+		klog.Warningf("invalid list volume req: %v", req)
+		return nil, err
+	}
+
+	//validate starting token if present
+	if len(req.GetStartingToken()) > 0 {
+		i, parseErr := strconv.ParseUint(req.StartingToken, 10, 32)
+		if parseErr != nil {
+			return nil, status.Errorf(codes.Aborted, "invalid starting token %s", parseErr.Error())
+		}
+		//check starting Token is greater than list of rbd volumes
+		if len(rbdVolumes) < int(i) {
+			return nil, status.Errorf(codes.Aborted, "invalid starting token %s", parseErr.Error())
+		}
+	}
+
+	var entries []*csi.ListVolumesResponse_Entry
+
+	for _, vol := range rbdVolumes {
```

Comment:
shouldn't the loop start from the starting token?

Reply:
`rbdVolumes` is a map any idea how to do this one?

I can do some hack like skip appending  if the counter value is less than startingToken

```
keys := make([]int, 0)
for k, _ := range rbdVolumes {
    keys = append(keys, k)
}
sort.Ints(keys)
for index, k := range keys {
    if index <i {
        continue
     }
               entries = append(entries, &csi.ListVolumesResponse_Entry{
		Volume: &csi.Volume{
			VolumeId:      rbdVolumes[k].VolID,
			CapacityBytes:  rbdVolumes[k].VolSize,
			VolumeContext: extractStoredVolOpt( rbdVolumes[k]),
			},
		})
}
```

@rootfs let me know your thoughts on this?

Reply:
looping over a map is not guaranteed to be ordered.

How about sorting the volume ID to make an ordered list?

Reply:
from the sample i pasted above does `sort.Ints(keys)` sort the keys but will test  it out

Reply:
done PTAL